src/adaptors/mod.rs

```diff
@@ -812,6 +812,72 @@ impl_tuple_combination!(Tuple10Combination Tuple9Combination; a b c d e f g h i)
 impl_tuple_combination!(Tuple11Combination Tuple10Combination; a b c d e f g h i j);
 impl_tuple_combination!(Tuple12Combination Tuple11Combination; a b c d e f g h i j k);
 
+#[derive(Debug, Clone)]
+pub struct ArrayCombinations<T, const R: usize> {
+    slice: Vec<T>,
+    indices: [usize; R],
+}
+
+impl<T, const R: usize> ArrayCombinations<T, R> {
+    pub fn new(slice: Vec<T>) -> Self {
+        debug_assert!(slice.len() >= R);
+
+        let mut indices = [0; R];
+        for i in 1..R {
+            indices[i] = i;
+        }
+
+        Self {
+            slice,
+            indices,
+        }
+    }
+}
+
+impl<T: Clone, const R: usize> Iterator for ArrayCombinations<T, R> {
+    type Item = [T; R];
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.indices[R-1] == self.slice.len() {
+            return None;
+        }
+
+        // SAFETY: uninitialized data is never read
+        let output = unsafe {
+            let mut output: [T; R] = std::mem::uninitialized();
+            for i in 0..R {
+                output[i] = self.slice[self.indices[i]].clone();
+            }
+            output
+        };
```

Comment:
I think the `SAFETY` note here is incorrect. When you overwrite `output[i]`, the value stored there previously will be dropped. But, in this case, that's probably _not_ a valid instance of `T`—it's uninitialized memory. I believe you need to use [`ptr::write`](https://doc.rust-lang.org/stable/std/ptr/fn.write.html) here.

And, instead of `mem::uninitialized`, use `MaybeUninit`. There's [some documentation on it about initializing arrays element-by-element]( https://doc.rust-lang.org/std/mem/union.MaybeUninit.html#initializing-an-array-element-by-element), though I expect the `mem::transmute` at the end won't work. You can use this as a work-around:
```rust
unsafe fn transmute<Src, Dst>(src: Src) -> Dst {
    use core::mem::ManuallyDrop;

    #[repr(C)]
    union Transmute<Src, Dst> {
        src: ManuallyDrop<Src>,
        dst: ManuallyDrop<Dst>,
    }

    ManuallyDrop::into_inner(Transmute { src: ManuallyDrop::new(src) }.dst)
}
```

Reply:
For initializing the array I wouldn't blindly copy that code because it will leak if `clone` panics. Instead I would use a guard struct that keeps track of how many items are initialized and drops them in case of panic.

For transmuting why not copy from [`MaybeUninit::array_assume_init](https://github.com/rust-lang/rust/blob/ef0ec303fa5ff6fa5654c7ea94322f3f9881f6c2/library/core/src/mem/maybe_uninit.rs#L869-L879)`?

Reply:
Good points about the dropping. Funnily enough I happened to read an article about the unsafety of array initialisation on HN earlier today.

Reply:
Read more into that thread. I found this crate https://crates.io/crates/array_init which could handle the safety of initialising the array. 

Similar functionality may make it into the std lib at some point https://github.com/rust-lang/rust/pull/75644